### ISSUE TITLE
Enqueuing jobs works too slow

### ISSUE THREAD

**Issue by bulbulhas** [sergeyzwezdin/Hangfire.Mongo#244]
Hi,
We are encountering some problems about our hangfire mongo implementation. We are really stuck at this point. I will try to explain our situation as much as i can, and maybe you can point me to right direction to find out what is our problem. 

We are using: 
- Hangfire 1.7.8
- Hangfire.Mongo 0,6,6

We have one storage, one dashboard and two different domain jobs. We are using scheduled jobs mostly. I will try to explain more about our structure.

- We have one project acts as dashboard, client and server. I ll call it project1 (client1 server1).
- We have second project in different domain acts as only server, call it server2
- A common library that contains the interfaces of server2 jobs. Referenced in both server2 and project1(for dashboard)
- Server1 has its own job methods and they are not in the common library. Server1 jobs are like total 4k scheduled jobs across 8 month interval.
- Server2 jobs was similar to this count until our new job implementation
- In our production environment, two instance of project1 and one instance of server2 is running
- Server2 and project1 worker counts are 1

Now problem: We added a new job for our server2 and job count in our application jumped to **2m**.
After this point we observed some jobs are not running., stayed on scheduled. After more investigation, we saw that job enqueue is became too slow. Servers trying to enqueue jobs scheduled a week ago. We are processing with 1 week delay. In order to pinpoint the problem, i tried to minify problem. Stoped all of our application. Created an app acts as server and dashboard, only activated **DelayedJobScheduler**. Added all job references and investigated console logs. 

After that i observed that job enqueue for this new application is really slow. 1 job would enqueued for 30 second-1min interval. We tried to tweak configuration options like queue polling, schedule polling interval but nothing helped. And this is where we are stuck at. what insights can you give me in order to solve our problem? 
I may not have been able to explain the problem enough. I will try to explain your questions in order to make sense of the problem in more detail.

**Comment by gottscj** [comment]
@bulbulhas,

Thanks for the thorough explanation. 
Would you be able to add your test app as PR?
What version of MongoDB are you using?

Thanks

**Comment by bulbulhas** [comment]
@gottscj 

Thanks for quick reply. I can provide code much faster here because it is relatively simple.
Our mongo cluster has 1 primary 3 secondary 1 arbitary nodes, 8 core and 16gb ram each
Mongo version : 4.2.1



	using Hangfire.Mongo;
	using System;
	using System.Collections.Generic;
	using Hangfire.Logging;
	using MongoDB.Driver;

	namespace Hangfire.Server
	{
		class Program
		{
			static void Main(string[] args)
			{
				var migrationOptions = new MongoMigrationOptions
				{
					Strategy = MongoMigrationStrategy.Migrate,
					BackupStrategy = MongoBackupStrategy.Collections
				};
				var storageOptions = new MongoStorageOptions
				{
					MigrationOptions = migrationOptions,
					QueuePollInterval = TimeSpan.FromSeconds(1)

				};

				var settings = MongoClientSettings.FromUrl(MongoUrl.Create("connectionString"));
				settings.MaxConnectionPoolSize = 1;
				settings.MaxConnectionIdleTime = TimeSpan.FromMinutes(1);
				settings.ConnectTimeout = TimeSpan.FromSeconds(55);
				settings.SocketTimeout = TimeSpan.FromSeconds(55);
				var mongoClient =  new MongoClient(settings);
				GlobalConfiguration.Configuration.UseColouredConsoleLogProvider(LogLevel.Trace);

				var mongoStorage = new MongoStorage(
					mongoClient,
					"databaseName",
					storageOptions);

				var processes = new List<IBackgroundProcess>();
				processes.Add(new DelayedJobScheduler());

				using (var server = new BackgroundProcessingServer(mongoStorage, processes))
				{
					Console.WriteLine("Custom BackgroundServer started. Press ENTER to exit...");
					Console.ReadLine();
				}
			}
		}
	}

**Comment by bulbulhas** [comment]
@gottscj Hi again,

With further investigation we saw that below code was the issue for our problem. In query builder "Score" field was the cause of our problem. After defining an index to it in our mongo database, delayed job scheduler started to process much much faster. I'll be creating a pr for this issue in the migration steps. If needed, I can fix the necessary parts in the code with your guidance.
```
 public override string GetFirstByLowestScoreFromSet(string key, double fromScore, double toScore)
 {
            if (key == null)
            {
                throw new ArgumentNullException(nameof(key));
            }

            if (toScore < fromScore)
            {
                throw new ArgumentException("The `toScore` value must be higher or equal to the `fromScore` value.");
            }
            
            return _dbContext
                .JobGraph
                .OfType<SetDto>()
                .Find(Builders<SetDto>.Filter.Regex(_ => _.Key, $"^{Regex.Escape(key)}") &
                      Builders<SetDto>.Filter.Gte(_ => _.Score, fromScore) &
                      Builders<SetDto>.Filter.Lte(_ => _.Score, toScore))
                .SortBy(_ => _.Score)
                .Project(_ => _.Value)
                .FirstOrDefault();
  }
```

**Comment by gottscj** [comment]
fixed in v0.7.12
